### PR TITLE
Use mutex to protect costmap reads.

### PR DESCRIPTION
**Pull request by LeifHookedWireless**
---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | No tickets known |
| Primary OS tested on | Ubuntu |
| Robotic platform tested on | Peanut's robot (sim) |

---

## Description of contribution in a few bullet points
* Protect costmap reads with the existing mutex.
* Otherwise costmap can be read during a map update and return 0.

## Description of documentation updates required from your changes

* No documentation changes needed.

---

## Future work that may be required in bullet points
* None expected.

#### For Maintainers: 
- [ ] Check that any new parameters added are updated in navigation.ros.org
- [ ] Check that any significant change is added to the migration guide
- [ ] Check that any new features **OR** changes to existing behaviors are reflected in the tuning guide
- [ ] Check that any new functions have Doxygen added
- [ ] Check that any new features have test coverage
- [ ] Check that any new plugins is added to the plugins page
- [ ] If BT Node, Additionally: add to BT's XML index of nodes for groot, BT package's readme table, and BT library lists

### CODE REVIEW

**Comment by SteveMacenski** [comment]
Where is this a problem, specifically. That's a lot of overhead to obtain a mutex when we call get/set costs thousands of times a loop. I would not merge this without an explicitly application example.

**Comment by LeifHookedWireless** [comment]
A specific example would be:
ObstaclesCritic::score()
ObstaclesCritic::costAtPose()
CollisionChecker::costAtPose()
Costmap2D::getCost()

which can occur simultaneously with

LayeredCostmap::updateMap()
which modifies Costmap2D in several places.

The result I see is that costAtPose() frequently returns 0 cost, in space that is non-zero and possibly even lethal.

**Comment by LeifHookedWireless** [comment]
Given how often getCost() is used, I suspect this is a problem in many places.
It's going to be tricky to optimize for the same reason, but I will take a look.

**Comment by SteveMacenski** [comment]
@LeifHookedWireless 

The algorithms lock the layered costmap's mutex when starting to plan: https://github.com/ros-planning/navigation2/blob/461a7ba1d296ef7224dde10d3e314a8f8d46fb81/nav2_smac_planner/src/smac_planner_hybrid.cpp#L316

Thus, updateMap is blocked: https://github.com/ros-planning/navigation2/blob/461a7ba1d296ef7224dde10d3e314a8f8d46fb81/nav2_costmap_2d/src/layered_costmap.cpp#L136

**Comment by LeifHookedWireless** [comment]
Sure, locking the update is half of it.
Locking the read is the (missing) other half.

**Comment by LeifHookedWireless** [comment]
So you're saying you want the costmap locked at a higher level?
Maybe in ObstaclesCritic::score()?
That will fix the case that I am aware of.

Personally, I would make the sure the reads are blocked everywhere, so that every call to getCost is protected, current and future.

**Comment by SteveMacenski** [comment]
I'm saying that the master costmap's mutex is already locked at the start of all of the planner/controller algorithms, for example as I linked above

**Comment by LeifHookedWireless** [comment]
I think that's true for all of the planners, and I see a lock in the DWB controller.
I don't see it in the MPPI controller, which is the one I'm trying to fix.
I think you're saying you want the lock in MPPIController::computeVelocityCommands()

FWIW, I don't see in a costmap lock in the RPP or RotationShim controllers either.

**Comment by LeifHookedWireless** [comment]
I locked the costmap in the MPPI controller.
That seems to fix the problem.

File: nav2_mppi_controller/src/controller.cpp [hunk 1]
```diff
@@ -83,9 +83,12 @@ geometry_msgs::msg::TwistStamped MPPIController::computeVelocityCommands(
   auto start = std::chrono::system_clock::now();
 #endif
 
-  std::lock_guard<std::mutex> lock(*parameters_handler_->getLock());
+  std::lock_guard<std::mutex> param_lock(*parameters_handler_->getLock());
   nav_msgs::msg::Path transformed_plan = path_handler_.transformPath(robot_pose);
 
+  nav2_costmap_2d::Costmap2D * costmap = costmap_ros_->getCostmap();
+  std::unique_lock<nav2_costmap_2d::Costmap2D::mutex_t> cosmtap_lock(*(costmap->getMutex()));
```
Comment (SteveMacenski) [inline nav2_mppi_controller/src/controller.cpp:90]:
```suggestion
  std::unique_lock<nav2_costmap_2d::Costmap2D::mutex_t> costmap_lock(*(costmap->getMutex()));
```

Comment (SteveMacenski) [reply nav2_mppi_controller/src/controller.cpp:90]: Your branch appears to be protected so I can't commit in this renaming fix. Please commit this in and I'll merge!

**Comment by SteveMacenski** [comment]
>  RPP or RotationShim controllers either

Can you add?

**Review by SteveMacenski — APPROVED** [review]

**Comment by SteveMacenski** [comment]
Ignore the linting failures in CI, apparently OSRF must have updated the profile. 

If you get that lock into the other algos, I can merge this in and backport immediately

**Comment by BriceRenaudeau** [comment]
Thanks for this fix, I was also wondering why I get 0 cost sometimes. :+1:

**Comment by LeifHookedWireless** [comment]
>https://github.com/ros-planning/navigation2/blob/461a7ba1d296ef7224dde10d3e314a8f8d46fb81/nav2_smac_planner/src/smac_planner_hybrid.cpp#L316
Shouldn't the planners and controllers use `std::shared_lock()`?
That way they both have read access to the costmap and won't block each other.
I guess we need to compile with C++17 for that...

**Comment by LeifHookedWireless** [comment]
I verified that RotationShim controller also had the issue and that adding the costmap lock fixes it.
I also added the lock to RPP controller, as requested, though I haven't had time to test it.
Will revisit this after ROSCon!

**Review by SteveMacenski — APPROVED** [review]

**Comment by LeifHookedWireless** [comment]
> I also added the lock to RPP controller, as requested, though I haven't had time to test it.
Confirmed that the bug was present in RPP (though not nearly as bad).
This PR does seem to fix it.

**Pull request merged**